pf/internal/logging/logging.go

```diff
@@ -0,0 +1,213 @@
+	ctx = setupRootLoggers(ctx, newLogSinkWriter(ctx, opts.LogSink))
+
+	if opts.URN != "" {
+		ctx = tflog.SetField(ctx, "urn", string(opts.URN))
+	}
+
+	if opts.ProviderName != "" {
+		ctx = tflog.SetField(ctx, "provider", opts.ProviderName)
+	}
+
+	if opts.ProviderVersion != "" {
+		ctx = tflog.SetField(ctx, "providerVersion", opts.ProviderVersion)
+	}
+
+	return ctx
+}
+
+// See InitLogging.
+type LogOptions struct {
+	LogSink         LogSink
+	ProviderName    string
+	ProviderVersion string
+	URN             resource.URN
+}
+
+// Abstracts the logging interface to HostClient. This is the interface providers use to report logging information back
+// to the Pulumi CLI over gRPC.
+type LogSink interface {
+	Log(context context.Context, sev diag.Severity, urn resource.URN, msg string) error
+}
+
+var _ LogSink = (*rprovider.HostClient)(nil)
+
+// Directs any logs written using the tflog API in the given Context to the given output.
+func setupRootLoggers(ctx context.Context, output io.Writer) context.Context {
+	desiredLevel := parseTfLogEnvVar()
+	sdkLoggerOptions := makeLoggerOptions("sdk", desiredLevel, output)
+	ctx = context.WithValue(ctx, sdkKey, hclog.New(sdkLoggerOptions))
+	ctx = context.WithValue(ctx, sdkOptionsKey, sdkLoggerOptions)
+	providerLoggerOptions := makeLoggerOptions("provider", desiredLevel, output)
+	ctx = context.WithValue(ctx, providerKey, hclog.New(providerLoggerOptions))
+	ctx = context.WithValue(ctx, providerOptionsKey, providerLoggerOptions)
+	return ctx
+}
+
+func makeLoggerOptions(name string, level hclog.Level, output io.Writer) *hclog.LoggerOptions {
+	if level == hclog.NoLevel {
+		level = hclog.Info
+	}
+	return &hclog.LoggerOptions{
+		Name:              name,
+		Output:            output,
+		Level:             level,
+		IndependentLevels: true,
+		IncludeLocation:   true,
+
+		// Empirically the value of 1 seems to work in the current Pulumi setup, @caller field now points to the
+		// file where the logging originates.
+		AdditionalLocationOffset: 1,
+	}
+}
+
+// Re-interprets strucutred JSON logs as calls against LogSink. To be used with SetupRootLoggers.
+func newLogSinkWriter(ctx context.Context, sink LogSink) io.Writer {
+	desiredLevel := parseTfLogEnvVar()
+	return &logSinkWriter{
+		desiredLevel: desiredLevel,
+		ctx:          ctx,
+		sink:         sink,
+	}
+}
+
+type logSinkWriter struct {
+	desiredLevel hclog.Level
+	ctx          context.Context
+	sink         LogSink
+}
+
+var _ io.Writer = &logSinkWriter{}
+
+func (w *logSinkWriter) Write(p []byte) (n int, err error) {
+	n = len(p)
+	if w.sink == nil {
+		return
+	}
+
+	raw := string(p)
+	level, raw := parseLevelFromRawString(raw)
+
+	if level <= w.desiredLevel {
+		return
+	}
+
+	urn, raw := parseUrnFromRawString(raw)
+	severity := logLevelToSeverity(level)
+
+	err = w.sink.Log(w.ctx, severity, urn, raw)
+	return
+}
+
+func parseTfLogEnvVar() hclog.Level {
+	env := os.Getenv("TF_LOG")
+	if env == "" {
```

Comment:
I think this reads better, but your free to discard.
```suggestion
	env, present := os.LookupEnv("TF_LOG")
	if !present {
```

You could also just remove the `env == ""` check, letting `default:` take care of it.